### PR TITLE
Add content headers to script headers object

### LINKED ISSUE #1314: Content-Type: Not passed in the Request Headers
_From @securityvoid on February 14, 2017 16:44_

Hello,
      It does not appear that "Content-Type" is handled so that it shows up under:
`req.headers`

This appears to be true for both Azure Functions Cli and the actual Azure deployments. (Not sure if this means I should open this issue elsewhere).

Interesting this data does appear to be within:
`context.bindingData`

I can parse this out, but can someone please explain why Content-Type wouldn't be passed with the rest of the headers?

I've got a form-data object that I need to parse to extract a file upload. There are packages that will do this automatically (e.g. busboy), if I pass in the Content-Type/boundary.

It seems weird to me that this wouldn't be included in the details.

_Copied from original issue: Azure/azure-functions-cli#68_

### PR DESCRIPTION
**Pull request by mamaso**
resolves #1314

### CODE REVIEW

File: test/WebJobs.Script.Tests.Integration/NodeEndToEndTests.cs [hunk 1]
```diff
@@ -752,6 +753,7 @@ public void ErrorFunction_DoesNotBringDownHost()
             Assert.Equal(testData, (string)resultObject["reqBody"]);
             Assert.Equal("string", (string)resultObject["reqRawBodyType"]);
             Assert.Equal(testData, (string)resultObject["reqRawBody"]);
+            Assert.Equal("text/plain", resultObject["reqHeaders"]["content-type"]);
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests.Integration/NodeEndToEndTests.cs]: Do we have other tests passing multi-value headers, headers with quality, etc.? Would be good to make sure those are handled as expected.

File: src/WebJobs.Script/Extensions/HttpRequestMessageExtensions.cs [hunk 2]
```diff
@@ -70,20 +70,8 @@ public static TValue GetRequestPropertyOrDefault<TValue>(this HttpRequestMessage
...
-            {
-                int idx = header.IndexOf(':');
-                string name = header.Substring(0, idx);
-                string value = header.Substring(idx + 1).Trim();
-                headers.Add(name, value);
-            }
-
-            return headers;
+            var headers = request.Content?.Headers.Concat(request.Headers) ?? request.Headers;
+            return headers.ToDictionary(p => p.Key, p => string.Join(" ", p.Value));
```
Comment (fabiocav) [inline src/WebJobs.Script/Extensions/HttpRequestMessageExtensions.cs]: Why using an empty space and not a comma here? This is a bit confusing. How is this being interpreted and why are we not using a comma?

Comment (mamaso) [inline src/WebJobs.Script/Extensions/HttpRequestMessageExtensions.cs]: Great question, after looking at [this httpheaders comment](https://github.com/dotnet/corefx/blob/bffef76f6af208e2042a2f27bc081ee908bb390b/src/System.Net.Http/src/System/Net/Http/Headers/ProductInfoHeaderParser.cs#L13) I'm going to go back to the original serializing code.  Not worth it trying to simplify if there are weird edge cases like this.

File: test/WebJobs.Script.Tests.Integration/NodeEndToEndTests.cs (hunk 1 above)
Comment (mamaso) [inline test/WebJobs.Script.Tests.Integration/NodeEndToEndTests.cs]: Another great comment, all of our tests use `User-Agent`, which as mentioned in the other comment is its own special snowflake as it uses `" "` separator. I'll add some tests with the accept header which uses standard comma separators.

File: src/WebJobs.Script/Extensions/HttpRequestMessageExtensions.cs [hunk 3]
```diff
@@ -70,9 +70,8 @@ public static TValue GetRequestPropertyOrDefault<TValue>(this HttpRequestMessage
 
         public static IDictionary<string, string> GetRawHeaders(this HttpRequestMessage request)
         {
-            Dictionary<string, string> headers = new Dictionary<string, string>();
-
-            var allHeadersRaw = request.Headers.ToString();
+            var headers = new Dictionary<string, string>();
+            var allHeadersRaw = request.Headers.ToString() + request.Content?.Headers?.ToString();
```
Comment (mathewc) [inline src/WebJobs.Script/Extensions/HttpRequestMessageExtensions.cs]: When concatenating these to ToString() values, is the first guaranteed to have a separator so they don't munge together?

File: src/WebJobs.Script/Extensions/HttpRequestMessageExtensions.cs [hunk 4]
```diff
@@ -70,9 +70,8 @@ public static TValue GetRequestPropertyOrDefault<TValue>(this HttpRequestMessage
...
             var rawHeaderLines = allHeadersRaw.Split(new string[] { Environment.NewLine }, StringSplitOptions.RemoveEmptyEntries);
```
Comment (mathewc) [inline src/WebJobs.Script/Extensions/HttpRequestMessageExtensions.cs]: The accept header is separated by commas + space - those are all parsed out the header collection ToString?

File: src/WebJobs.Script/Extensions/HttpRequestMessageExtensions.cs [hunk 5]
```diff
@@ -70,9 +70,8 @@ public static TValue GetRequestPropertyOrDefault<TValue>(this HttpRequestMessage
 
         public static IDictionary<string, string> GetRawHeaders(this HttpRequestMessage request)
         {
-            Dictionary<string, string> headers = new Dictionary<string, string>();
-
-            var allHeadersRaw = request.Headers.ToString();
+            var headers = new Dictionary<string, string>();
```
Comment (mathewc) [inline src/WebJobs.Script/Extensions/HttpRequestMessageExtensions.cs]: We have logic here https://github.com/Azure/azure-webjobs-sdk-script/blob/658d4d0c9dce9af940bd1a4fd8d5fb0f2fb85e90/src/WebJobs.Script/Binding/HttpBinding.cs#L290 that goes the other way - takes top level headers and applies them to a response. So what you're doing makes this symmetric for request content.

File: test/WebJobs.Script.Tests.Integration/NodeEndToEndTests.cs [hunk 6]
```diff
@@ -736,6 +739,7 @@ public void ErrorFunction_DoesNotBringDownHost()
                 Content = new StringContent(testData)
             };
             request.SetConfiguration(new HttpConfiguration());
+            request.Content.Headers.ContentType = new MediaTypeHeaderValue("text/plain");
```
Comment (mathewc) [inline test/WebJobs.Script.Tests.Integration/NodeEndToEndTests.cs]: Can you add like 2 more content headers and verify them below to test the multiplicity? I.e. symmetric with https://github.com/Azure/azure-webjobs-sdk-script/blob/658d4d0c9dce9af940bd1a4fd8d5fb0f2fb85e90/src/WebJobs.Script/Binding/HttpBinding.cs#L290

File: src/WebJobs.Script/Extensions/HttpRequestMessageExtensions.cs (hunk 3 above)
Comment (mamaso) [inline src/WebJobs.Script/Extensions/HttpRequestMessageExtensions.cs]: will add a newline separator to ensure that the headers are split

**Comment by SteveALee** [comment]
Thanks!
